flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/ReducingUpsertSink.java

```diff
@@ -43,19 +45,19 @@
     private final DataType physicalDataType;
     private final int[] keyProjection;
     private final SinkBufferFlushMode bufferFlushMode;
-    private final Function<RowData, RowData> valueCopyFunction;
+    @Nullable private final TypeSerializer<RowData> typeSerializer;
 
     ReducingUpsertSink(
             Sink<RowData, ?, WriterState, ?> wrappedSink,
             DataType physicalDataType,
             int[] keyProjection,
             SinkBufferFlushMode bufferFlushMode,
-            Function<RowData, RowData> valueCopyFunction) {
```

Comment:
Here we could also use `SerializableFunction` but your change is also good (in general not a huge fan of nullable fields that imply different behavior).

Reply:
I think making the function serializable did not really work because of the object reuse that the sink is using ...

Reply:
@fapaul By using the TypeSerializer is against what you mean **making the function serializable did not really work**?